### ISSUE TITLE
Debugging user issues better

### ISSUE THREAD

**Issue by xoriole** [Tribler/tribler#4930]
Some bugs on Tribler are difficult to fix because they are difficult to reproduce. Knowing the internal state of Tribler during the faulty execution would be immensely useful for debugging such issue. 
Some directions discussed at today's meeting for that:
- Full state recording, extreme logging in Tribler/python. User can then upload the logs along with the error report. 
- Debug option in PyInstaller build

Seems there is a related old ticket [#2910](https://github.com/Tribler/tribler/issues/2910)

**Comment by qstokkink** [comment]
We already have the `--debug` flag. What do you propose to extend this with?

**Comment by ichorid** [comment]
The problem is, if you need to turn on the "debugging mode" to capture the complete Tribler state, you'll have to know *in advance* that Tribler is going to crash. This will only help the most advanced and determined user.

An alternative route is to do it the way all major OSes do it: in case of crash dump the whole process memory state on disk, *replacing the previous dump*. Then, the reported can have a checkbox to attach the dump. Note, though, that this kind of stuff will invariably leak private user info... Even if we try to filter the dump for things like, e.g. private key, before sending.

**Comment by xoriole** [comment]
@qstokkink Yes, we already have the flag but it may not be sufficient to reproduce the issues. Besides, we don't get the user logs in the error report. So, I think we should extend the error reporter to add a choice (disabled by default) whether to upload the logs along with the report.
@ichorid crash dump and attaching to the error report is a good idea and I'm in favor of it. Regarding privacy, it is a user consent to attach the files. Also, we could try some basic anonymisation on data.

**Comment by qstokkink** [comment]
Two notes on crash dumps:
 - They are big (our server is already filling up) and can take in the order of minutes to properly capture.
 - Someone has to plow through them to find out what was wrong (which might be OS specific).

Regarding logging:
 - Three years ago I performed a clean performance experiment on Dispersy throughput. By stripping out all logging I gained 400 MB/s in download speed. Be careful with overzealous logging, especially in Python - which is basically single thread execution.

**Comment by ichorid** [comment]
relate to _the plan_ #5348
